### ISSUE TITLE
gradle/gradle build relies on disallowed methods from lazy configuration actions

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle-native#864]
For our performance testing plugins, we attempt to copy the inputs from any configured sample tasks to the performance test task.

This means we're doing something like (this is simplified, what we do is more complicated):
```
tasks.withType(SampleTask).configureEach { sample ->
     tasks.withType(PerformanceTest).configureEach { test ->
          test.inputs.properties.addAll(sample.inputs.properties)
     }
}
```

We no longer allow this kind of nesting because it can cause behavior that's specific to which tasks were realized in the build.  In our case, this is exactly what we want (only configure PerformanceTest if a SampleTask was realized).

I've rewritten the build logic a little to not depend on this, but it's still horrible.

**Comment by lacasseio** [comment]
What do we think about using the `whenElementKnown(Action<? extends Provider>)` pattern as discussed [here](https://gradle-community.slack.com/archives/CDB9Q0MPT/p1539278088000100)? It would give the material to address this issue:
```
tasks.withType(SampleTask).whenElementKnow { TaskProvider<SampleTask> sample ->
     tasks.withType(PerformanceTest).configureEach { PerformanceTest test ->
          test.inputs.properties.addAll(sample.get().inputs.properties)  // We can also map `sample` to the property
     }
}
```

Then `whenElementKnow` would fire when the element is added or created or register. But it would be only available through a provider that we shouldn't allow to query at that time.

**Comment by big-guy** [comment]
I haven't looked deeply at this again yet, but the use case is pretty strange IMO.  It's something like "if this task is _realized_, then this other task is configured differently".

I was thinking we could solve this by putting all the generated samples into say `build/samples/` and adding that as an input to all `PerformanceTest`s.  I think that models the relationship a bit better (the samples are an input to the test).  I think we're doing some of this funkiness because we were trying to make the PerformanceTests cacheable.

**Comment by lacasseio** [comment]
I wrote a fix for this issue in this PR: https://github.com/gradle/gradle/pull/7406

I'm unsure how to test the parity of this PR with the previous code. I spend too much time trying to run a performance test while receiving various compile errors that end up not been errors but just fail import from Idea which doesn't go away. @big-guy do you have a good idea how to test the code?

**Comment by big-guy** [comment]
I think we want to try to get rid of the whole "configure performance tests with sample generator properties" step.  I think this is somewhat what @wolfs was trying to do: https://github.com/gradle/gradle/pull/6600

I think we should do something like:
- Remove the code that adds the inputs of a sample task to the performance tests
- Generate samples into `performance/build/samples/<whatever>`
- Make `prepareSamples` a no-op and/or remove it entirely.
- Change the performance tests to assert that they can find their samples (if they don't already do this) and recommend running `gradle <sample name>` when it cannot be found.
- Add `performance/build/samples` as an input to all PerformanceTest
- Make sure there's no dependency between any sample and a PerformanceTest

It looks like a lot of that is in https://github.com/gradle/gradle/pull/6600

So instead of generating all possible samples, we rely on the PerformanceTest to fail if the sample it needs is not available.  This may mean we will need to clean `performance/build/samples` before performance builds, if we're not already doing that.

**Comment by wolfs** [comment]
@big-guy I would favor https://github.com/gradle/gradle-private/issues/1481 over my experiment in gradle/gradle#6600: Then all template generator tasks would be `JavaExec` like tasks, which has various advantages:
- The only input to track is the classpath of the template generator project. If there are no changes, then the templates are the same.
- The tests could generate the templates on demand, without relying on other tasks to pre-create them for the test
- It is not necessary to generate all the performance templates only to snapshot the actual projects as inputs to the distributed performance test.

**Comment by lacasseio** [comment]
I agree it makes sense to improve it in this way. However, my question would be is the change in the PR good enough for now or do we want to spend the time to fix it how it is explained in the comments?

I'm coming more from the side of I don't know how this part of the code base works and would need to move cautiously not to break anything as testing this code isn't the easiest. It would take at least the rest of the week to do the fix as explained. Do we need the proper fix now or can we live with just removing the deprecation warning with an equivalent code?

**Comment by wolfs** [comment]
@lacasseio I wouldn't do https://github.com/gradle/gradle/pull/6600 if we don't improve caching for the templates so the distribution build would be really fast. Since that also will take some time (including tracking down cache misses), I would prefer to invest the time in https://github.com/gradle/gradle-private/issues/1481.

I have currently no objections to leave the logic as is. How expensive would it be to realize all the template tasks all the time? That would also make the current solution easier, right?

**Comment by lacasseio** [comment]
The change in term of performance is a no-op compared to the previous code. It simply realizes the task when the inputs of the performance task are been snapshotted. The fix in the PR is moving the realization from during configuration to during task graph building and execution. We can invest time now to improve the situation but it seems out-of-scope of the current issue.

**Comment by big-guy** [comment]
Waiting for https://github.com/gradle/gradle/pull/7337 to get into master
